src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+  (h : is_cusp_form_of_weight_and_level k Γ f) : is_modular_form_of_weight_and_level k Γ f :=
+{ hol := h.1,
+  transf := h.2,
+  infinity := λ (A : SL(2,ℤ)), by {apply filter.zero_at_filter_is_bounded_at_filter _ _ (h.3 A)} }
+
+ /-- The zero modular form is a modular form-/
+lemma zero_mod_form : is_modular_form_of_weight_and_level k Γ 0 :=
+by apply_rules [is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level,
+    zero_cusp_form]
+
+
+/-- This is the space of modular forms of weight `k` and level `Γ`-/
+def space_of_mod_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := { f : ℍ → ℂ | is_modular_form_of_weight_and_level k Γ f },
+  zero_mem':= by { simp only [set.mem_set_of_eq], apply zero_mod_form},
+  add_mem' := begin intros a b ha hb,
+    split,
+    { exact mdifferentiable_add ha.hol hb.hol },
+    { exact (weakly_modular_submodule k Γ).add_mem' ha.transf hb.transf} ,
+    { intro A,
+      rw slash_action.add_action,
+      exact (bounded_at_im_infty_subalgebra ℂ).add_mem' (ha.infinity A) (hb.infinity A)}
+    end,
+  smul_mem' := begin intros c f hf,
+    split,
+    { exact mdifferentiable_smul _ hf.hol },
+    { exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf },
+    { intro A,
+      rw slash_action.smul_action,
+      apply ((bounded_at_im_infty_subalgebra ℂ).smul_mem (hf.infinity A))}
+    end}
+
+localized "notation `M`:= space_of_mod_forms_of_weight_and_level " in modular_forms
+
+/-- This is the space of cuspforms of weigth `k` and level `Γ` -/
+def space_of_cusp_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := is_cusp_form_of_weight_and_level k Γ,
+  zero_mem' := by apply zero_cusp_form,
+  add_mem' := begin intros a b ha hb,
+    split,
+    { exact mdifferentiable_add ha.hol hb.hol },
+    { exact (weakly_modular_submodule k Γ).add_mem' ha.transf hb.transf },
+    { intro A,
+      rw slash_action.add_action,
+      apply ((zero_at_im_infty_submodule ℂ).add_mem' (ha.infinity A) (hb.infinity A))}
+    end,
+  smul_mem' :=begin intros c f hf,
+    split,
+    { exact mdifferentiable_smul _ hf.hol },
+    { exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf },
+    { intro A,
+      rw slash_action.smul_action,
+      apply (zero_at_im_infty_submodule ℂ).smul_mem' c (hf.infinity A)},
+    end}
+
+localized "notation `S`:= space_of_cusp_forms_of_weight_and_level" in modular_forms
+
+/--The product of two modular forms is a modular form whose weight is the sum of the weights-/
+lemma mul_modform (k_1 k_2 : ℤ) (Γ : subgroup SL(2,ℤ)) (f g : ℍ → ℂ)
+  (hf : f ∈ M k_1 Γ) (hg : g ∈ M k_2 Γ) : f * g ∈ (M (k_1 + k_2) Γ) :=
+begin
+  refine ⟨mdifferentiable_mul hf.1 hg.1, mul_modular _ _ _ _ _ hf.2 hg.2, λ A, _⟩,
+  rw [slash_mul_SL2 k_1 k_2 A f g],
+  exact prod_of_bounded_is_bounded (hf.infinity A) (hg.infinity A),
+end
```

Comment:
Again, better to do the properties by name:
```suggestion
{ hol := mdifferentiable_mul hf.hol hg.hol,
  transf := mul_modular hf.transf hg.transf,
  infinity := λ A, by simpa [slash_mul_SL2] using prod_of_bounded_is_bounded (hf.infinity A) (hg.infinity A) }
```
I also recommend renaming `prod_of_bounded_is_bounded` to `is_bounded_at_im_infty.mul`: (1) the naming convention is `mul` not `prod`; (2) this will allow for dot notation in use cases like this one.